_posts/2023-12-20-standard-libraries.markdown

```diff
@@ -0,0 +1,407 @@
+to its corresponding character.
+But let's pretend we're writing Real Code for the moment so I can show off the `Std.Unicode` library;
+hey you never know, later puzzles might have proper UTF8 content!
+
+```dafny
+import Std.BoundedInts
+import Std.FileIO
+import Std.Unicode.UnicodeStringsWithUnicodeChar
+import Std.Wrappers
+
+method ReadPuzzleInput() returns (input: Wrappers.Result<string, string>) {
+  var bytesAsBVs :- FileIO.ReadBytesFromFile("input.txt");
+  
+  var bytes := seq(|bytesAsBVs|, i requires 0 <= i < |bytesAsBVs| => bytesAsBVs[i] as BoundedInts.uint8);
+  input := UnicodeStringsWithUnicodeChar.FromUTF8Checked(bytes).ToResult("Invalid UTF8");
+}
+```
+
+A few notes:
+
+* The `UnicodeStringsWithUnicodeChar` module is named that way to emphasize that it is only
+  correct to use with the `--unicode-char:true` mode, which is [on by default in Dafny 4.x](https://dafny.org/blog/2023/03/03/dafny-4-released/#unicode-strings). 
+* The `FromUTF8Checked` function takes in a `seq<uint8>`, where `uint8` is a `newtype` definition from `Std.BoundedInts`.
+  This library defines lots of common fixed-bit-width integer types that the Dafny compiler will map to
+  the native integer types of the target language for improved efficiency and memory usage.
+* Because `bv8` is a distinct type from `uint8`, we have to explicitly convert between them.
+  This is a side effect of the Dafny standard libraries having multiple independent contributions from various
+  projects, and you can expect future versions of Dafny to provide more versions of common utilities
+  to make combining things a bit smoother.
+* `FromUTF8Checked` returns an `Option<string>`, but `FileIO.ReadBytesFrom` returns a `Result<seq<bv8>, string>`.
+  To unify with a common result type, we use `Option.ToResult` to convert the former value to a `Result`,
+  which is a handy and common trick. See [the documentation for `Std.Wrappers`](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyStandardLibraries/src/Std/Wrappers.md) for more details.
+
+## Sprinting to the Finish
+
+Now let's put it all together and write our main method:
+
+```dafny
+method Main() {
+  var input :- expect ReadPuzzleInput();
+
+  var lines := Seq.Split(s, '\n');
+
+  var calibrationValues :- expect Seq.MapWithResult(CalibrationValue, lines);
+
+  var total := Seq.FoldLeft((x, y) => x + y, 0, calibrationValues);
+  print total, "\n";
+}
+```
+
+We've used one more trick with failure-compatible types in this method:
+`:- expect`, a variation on the elephant operator to make "assign or halt" statements.
+`ReadPuzzleInput` returns a `Result`, but if it fails there's nothing more we can do
+in the main method than printing an error and exiting.
+`:- expect` means that if the right-hand side of the statement is a failure,
+Dafny will immediately halt and print the failure value to the console.
+This is particularly valuable for writing tests in Dafny.
```

Comment:
I just discovered that tests can return failure-compatible types as well so you can use :- without expect. Not sure if it's worth mentioning here but perhaps in the future.

Reply:
Indeed I'm still debating whether to frame the solution a little differently. As soon as I added the CI check of the solution I thought "this should really be a `{:test}` instead of a Main method". But on the other hand in the puzzle solving context you don't actually know the answer so you do need a Main method you just run as well, plus I feel like printing out the answer is just more exciting than a passing test case. :)

In fact I think I'll actually call this out explicitly and do both, since I'm already calling out similar Real Code issues outside of the puzzle context anyway.